src/daft-parquet/src/metadata.rs

```diff
@@ -213,73 +216,109 @@ fn apply_field_ids_to_parquet_file_metadata(
-    io_client: Arc<IOClient>,
-    io_stats: Option<IOStatsRef>,
-    field_id_mapping: Option<Arc<BTreeMap<i32, Field>>>,
-) -> super::Result<FileMetaData> {
-    const FOOTER_SIZE: usize = 8;
+pub(crate) fn validate_footer_magic(uri: &str, buffer: &[u8]) -> super::Result<()> {
     const PARQUET_MAGIC: [u8; 4] = [b'P', b'A', b'R', b'1'];
-    if size < 12 {
+
+    if buffer.len() < FOOTER_SIZE {
         return Err(Error::FileTooSmall {
             path: uri.into(),
-            file_size: size,
+            file_size: buffer.len(),
         });
     }
 
-    /// The number of bytes read at the end of the parquet file on first read
-    const DEFAULT_FOOTER_READ_SIZE: usize = 128 * 1024;
-    let default_end_len = std::cmp::min(DEFAULT_FOOTER_READ_SIZE, size);
-
-    let start = size.saturating_sub(default_end_len);
-    let mut data = io_client
-        .single_url_get(uri.into(), Some(start..size), io_stats.clone())
-        .await?
-        .bytes()
-        .await?;
-
-    let buffer = data.as_ref();
     if buffer[buffer.len() - 4..] != PARQUET_MAGIC {
         return Err(Error::InvalidParquetFile {
             path: uri.into(),
             footer: buffer[buffer.len() - 4..].into(),
         });
     }
-    let metadata_size = metadata_len(buffer, default_end_len);
-    let footer_len = FOOTER_SIZE + metadata_size as usize;
+    Ok(())
+}
 
-    if footer_len > size {
-        return Err(Error::InvalidParquetFooterSize {
-            path: uri.into(),
-            footer_size: footer_len,
-            file_size: size,
-        });
+pub(crate) async fn read_parquet_metadata(
+    uri: &str,
+    file_size: Option<usize>,
+    io_client: Arc<IOClient>,
+    io_stats: Option<IOStatsRef>,
+    field_id_mapping: Option<Arc<BTreeMap<i32, Field>>>,
+    default_footer_read_size: Option<usize>,
+) -> super::Result<FileMetaData> {
+    async fn fetch_data(
+        io_client: Arc<IOClient>,
+        uri: &str,
+        range: GetRange,
+        io_stats: Option<IOStatsRef>,
+    ) -> daft_io::Result<Bytes> {
+        io_client
+            .single_url_get(uri.into(), Some(range), io_stats)
+            .await?
+            .bytes()
+            .await
+    }
+
+    // Check the minimum value of file size if provided
+    if let Some(size) = file_size {
+        if size < 12 {
+            return Err(Error::FileTooSmall {
+                path: uri.into(),
+                file_size: size,
+            });
+        }
+    }
+
+    /// The number of bytes read at the end of the parquet file on first read
+    const DEFAULT_FOOTER_READ_SIZE: usize = 128 * 1024;
+    let footer_read_size = default_footer_read_size
+        .unwrap_or(DEFAULT_FOOTER_READ_SIZE)
+        .max(FOOTER_SIZE);
+    let range = match file_size {
+        None => GetRange::Suffix(footer_read_size),
+        Some(size) => {
+            let default_end_len = std::cmp::min(footer_read_size, size);
+            let start = size.saturating_sub(default_end_len);
```

Comment:
Nit:
```suggestion
            let read_size = footer_read_size.min(size);
            let start = size - read_size;
```
saturating sub isn't needed since we already do a `min` on the lengths